relay-server/src/services/upload.rs

```diff
@@ -118,38 +65,143 @@ impl OutcomeError for Error {
 ///
 /// Accepts upload requests and maintains a list of concurrent uploads.
 pub struct UploadService {
-    pending_requests: FuturesUnordered<BoxFuture<'static, Result<(), Rejected<Error>>>>,
+    pending_requests: FuturesUnordered<BoxFuture<'static, ()>>,
     max_concurrent_requests: usize,
+    inner: Arc<UploadServiceInner>,
+}
+struct UploadServiceInner {
     timeout: Duration,
+    store: Addr<Store>,
+
+    objectstore_client: Client,
+    attachments_usecase: Usecase,
+}
+
+impl UploadServiceInner {
+    /// Spend a limited time trying to upload an attachment, and emit outcomes if this fails.
+    ///
+    /// Returns an [`UploadedAttachment`] if the upload was successful.
+    async fn handle_envelope(&self, mut envelope: TypedEnvelope<Processed>) -> () {
+        let scoping = envelope.scoping();
+        let session = self
+            .attachments_usecase
+            .for_project(scoping.organization_id.value(), scoping.project_id.value())
+            .session(&self.objectstore_client);
+
+        let attachments = envelope
+            .envelope_mut()
+            .items_mut()
+            .filter(|item| *item.ty() == ItemType::Attachment);
+        let upload_results = if let Ok(session) = session {
+            futures::future::join_all(
+                attachments.map(|attachment| self.handle_attachment(&session, attachment)),
+            )
+            .await
+        } else {
+            // if we have any kind of error constructing the upload session, we mark all attachments as failed
+            attachments.map(|_| Err(Error::UploadFailed)).collect()
+        };
+
+        // track the results of attachment upload
+        let managed_attachments = envelope
+            .envelope()
+            .items()
+            .filter(|item| *item.ty() == ItemType::Attachment)
+            .map(|item| Managed::from_envelope(&envelope, item));
+        for (attachment, result) in managed_attachments.zip(upload_results) {
+            let result_msg = match result {
+                Ok(()) => {
+                    attachment.accept(|_| ());
+                    "success"
+                }
+                Err(error) => attachment.reject_err(error).into_inner().as_str(),
```

Comment:
This reports a negative outcome for the attachment as if it was lost, but the attachment is still forwarded to the store service. For the event attachment use case I believe we should simply not make these uploaded attachments `Managed`.

For the span attachment use case we will want them to be `Managed`, because there will be no other copy of the data that can own the outcomes.

@Dav1dde thoughts?